Add support for rgw loadbalancer based on haproxy & keepalived

guihecheng opened the pull request:
This is a *basic* version of a loadbalancer solution for multiple rgw instances.
It includes:
- An infrastructure playbook for haproxy & keepalived setup -- rgw-loadbalancer.yml
- An infrastructure playbook for purging -- purge-rgw-loadbalancer.yml
- Roles definition for haproxy & keepalived

Any comment(s) would be highly appreciated :)

guihecheng commented:
hi @leseb, any comments on this one?

leseb commented on roles/ceph-haproxy/tasks/pre_requisite.yml:
```diff
@@ -0,0 +1,19 @@
+    - ansible_os_family == 'RedHat'
+
+- name: "generate haproxy configuration file: haproxy.cfg"
+  action: config_template
```
Please use default Ansible`template` instead.

guihecheng replied:
ok, I shall fix this.

dsavineau replied:
template module is a better idea because you could add file syntax validation via `validate: 'haproxy -f %s -c`
You should also add a handler to reload the config when config change. Currently any changes in the config won't be effective.

guihecheng replied:
@dsavineau thanks, I shall add validation and reloading.

leseb commented on infrastructure-playbooks/purge-rgw-loadbalancer.yml:
```diff
@@ -0,0 +1,113 @@
+    file:
+      path: "{{ item }}"
+      state: absent
+    with_items:
```
You don't need the loop since you have a single item.

guihecheng replied:
oh, right, I shall fix all the "single" loops.

leseb commented on infrastructure-playbooks/purge-rgw-loadbalancer.yml:
```diff
@@ -0,0 +1,113 @@
+      name: keepalived
+      state: stopped
+      enabled: no
+    failed_when: false
```
not needed

leseb commented on infrastructure-playbooks/purge-rgw-loadbalancer.yml:
```diff
@@ -0,0 +1,113 @@
+      state: stopped
+      enabled: no
+    failed_when: false
+    when: ansible_service_mgr == 'systemd'
```
service detects the init system automatically so you don't need that.

guihecheng replied:
ok, that's nice.

leseb commented on infrastructure-playbooks/purge-rgw-loadbalancer.yml:
```diff
@@ -0,0 +1,113 @@
+      name: haproxy
+      state: stopped
+      enabled: no
+    failed_when: false
```
ditto

leseb commented on infrastructure-playbooks/purge-rgw-loadbalancer.yml:
```diff
@@ -0,0 +1,113 @@
+    file:
+      path: "{{ item }}"
+      state: absent
+    with_items:
```
ditto

leseb commented on roles/ceph-haproxy/tasks/pre_requisite.yml:
```diff
@@ -0,0 +1,19 @@
+  package:
+    name: "{{ item }}"
+    state: present
+  with_items:
```
ditto

leseb commented on roles/ceph-haproxy/tasks/pre_requisite.yml:
```diff
@@ -0,0 +1,19 @@
+    state: present
+  with_items:
+    - haproxy
+  when:
```
why?

leseb commented on roles/ceph-keepalived/tasks/pre_requisite.yml:
```diff
@@ -0,0 +1,39 @@
+  package:
+    name: "{{ item }}"
+    state: present
+  with_items:
```
ditto

leseb commented on roles/ceph-keepalived/tasks/pre_requisite.yml:
```diff
@@ -0,0 +1,39 @@
+    state: present
+  with_items:
+    - keepalived
+  when:
```
ditto

leseb commented on roles/ceph-keepalived/tasks/pre_requisite.yml:
```diff
@@ -0,0 +1,39 @@
+      index_var: index
+
+- name: "generate keepalived: configuration file: keepalived.conf"
+  action: config_template
```
ditto

dsavineau replied:
add a handler there too

mattbenjamin commented:
@guihecheng interesting! adding a couple of reviewers

guihecheng commented:
@mattbenjamin oh, nice to see you here, matt, thanks for your help!
Actually, I've tried haproxy & keepalived in our production environment, but we are not experts on haproxy & keepalived,we really need all variable suggestions from all the reviewers :)

hoerup commented on roles/ceph-haproxy/templates/haproxy.cfg.j2:
```diff
@@ -0,0 +1,44 @@
+    default_backend rgw-backend
+
+backend rgw-backend
+    mode http
```
no need to set mode http when you have set it in defaults

guihecheng replied:
@hoerup OK, thanks.

dsavineau commented on roles/ceph-haproxy/meta/main.yml:
```diff
@@ -0,0 +1,13 @@
+  author: Gui Hecheng
+  description: Config HAProxy
+  license: Apache
+  min_ansible_version: 2.3
```
include_tasks have been introduced in ansible 2.4 so the min_ansible_version should reflect that.

guihecheng replied:
@dsavineau thanks for pointing the exact version out, and I shall add firewalld rules and selinux configs.

dsavineau commented on roles/ceph-keepalived/meta/main.yml:
```diff
@@ -0,0 +1,13 @@
+  author: Gui Hecheng
+  description: Config Keepalived
+  license: Apache
+  min_ansible_version: 2.3
```
ditto

dsavineau commented on infrastructure-playbooks/rgw-loadbalancer.yml:
```diff
@@ -0,0 +1,27 @@
+
+- hosts: haproxys
+  vars:
+    - rgw_group_name: rgws
```
why hardcoding the rgw_group_name here when you can reuse `"{{ rgw_group_name|default('rgws') }}"` like before ?

guihecheng replied:
@dsavineau yes, you are right.

dsavineau commented on roles/ceph-keepalived/tasks/pre_requisite.yml:
```diff
@@ -0,0 +1,39 @@
+    group: "root"
+    mode: "0644"
+
+- name: "set master for vrrp_instance: configuration file: keepalived.conf"
```
why this isn't part of the keepalived template ?

guihecheng replied:
@dsavineau because for now, I don't want to make the template file looks too complex by checking the MASTER/BACKUP state, but I shall try to put this part into the template file in the next version.

guihecheng commented:
hi all, I've updated a version with some changes:
- merge the extra playbook into site.yml, this mainly intends to let the option `configure_firewalld` to control the behavior of haproxy and keepalived firewall settings.
- add haproxy and keepalived firewall settings.
- merge haproxy and keepalived role into rgw-loadbalancer role

guihecheng commented:
@leseb hi, any interest to review the new version ?
There's quite a lot of changes~

guihecheng commented:
ping @leseb ?

hoerup commented:
maybe @guits  or @dsavineau  can have a look at the updated version ?

guits commented on infrastructure-playbooks/purge-rgw-loadbalancer.yml:
```diff
@@ -0,0 +1,76 @@
+---
```
why not including this in existing purge playbooks?

guihecheng replied:
@guits Emmm, that's a good idea, I shall put it into the `purge-cluster.yml` playbook, thanks.

guihecheng commented:
@guits merged the lb purging stuff into existing `purge-cluster.yml`, thanks.

guihecheng commented:
ping @guits @leseb

guits commented on roles/ceph-rgw-loadbalancer/meta/main.yml:
```diff
@@ -0,0 +1,13 @@
+  author: Gui Hecheng
+  description: Config HAProxy & Keepalived
+  license: Apache
+  min_ansible_version: 2.4
```
```suggestion
  min_ansible_version: 2.8
```

guihecheng replied:
@guits as I understand, version 2.4 is too low for some ansible features ?

guits replied:
stable-4.0 (which is the only branch where this patch will be backported) supports ansible 2.8 only

guihecheng replied:
OK

guits commented on roles/ceph-rgw-loadbalancer/defaults/main.yml:
```diff
@@ -0,0 +1,16 @@
+---
+# Variables here are applicable to all host groups NOT roles
+
+# This sample file generated by generate_group_vars_sample.sh
```
this comment shouldn't be there

guihecheng replied:
ok, will remove it.

guits replied:
do not forget to run the `./generate_group_vars_sample.sh` script

thanks!

guihecheng replied:
oh, that's a missing point, I shall use this facility.

guits commented:
@guihecheng we're almost there.

I know it's been a while, thanks for your patience!

guihecheng commented:
@guits thanks for your comments, I've adopted the changes, hopt it works this time :)

guits commented:
@guihecheng If I run `./generate_group_vars_sample.sh` in your branch I get this diff:

```diff
diff --git a/group_vars/rgwloadbalancers.yml.sample b/group_vars/rgwloadbalancers.yml.sample
index 76fab9f7..0784a2d8 100644
--- a/group_vars/rgwloadbalancers.yml.sample
+++ b/group_vars/rgwloadbalancers.yml.sample
@@ -14,10 +14,4 @@ dummy:
 ###########
 
 #haproxy_frontend_port: 80
-#
-#virtual_ips:
-#   - 192.168.238.250
-#   - 192.168.238.251
-#
-#virtual_ip_netmask: 24
-#virtual_ip_interface: ens33
+
```

guihecheng commented:
@guits yes, I added these samples by hand, because it is impossible to pick defaults values for these virtual ip information, users could only arrange their virtual ips by themselfs.
But here, I still want to give examples as a hint, so I added these.

guits commented:
@guihecheng so you can leave them commented out in `roles/ceph-rgw-loadbalancer/defaults/main.yml`
we must keep each sample file synced with its corresponding `defaults/main.yml`

guits commented on roles/ceph-rgw-loadbalancer/defaults/main.yml:
```diff
@@ -0,0 +1,8 @@
+# GENERAL #
+###########
+
+haproxy_frontend_port: 80
```
```suggestion
haproxy_frontend_port: 80
#
#virtual_ips:
#   - 192.168.238.250
#   - 192.168.238.251
#
#virtual_ip_netmask: 24
#virtual_ip_interface: ens33
```

then you just have to run `./generate_group_vars_sample.sh` and the diff should disappear in the corresponding sample file.

guihecheng replied:
this is a good way, thanks, I'll update soon.

guihecheng commented:
@guits updated, thanks~

guits commented:
not sure what's wrong with mergify, merging anyway.

secret104278 commented:
How do you handle ssl with haproxy deployed by this ansible role?

hoerup commented:
@secret104278  look at https://github.com/ceph/ceph-ansible/blob/master/group_vars/rgwloadbalancers.yml.sample#L18
